polaris-core/src/main/java/org/apache/polaris/core/persistence/resolver/Resolver.java

```diff
@@ -703,61 +693,73 @@ private ResolverStatus resolvePaths(
     }
 
-    // activate all principal roles which still exist
-    for (PolarisGrantRecord grantRecord : this.resolvedCallerPrincipal.getGrantRecordsAsGrantee()) {
-      if (grantRecord.getPrivilegeCode() == PolarisPrivilege.PRINCIPAL_ROLE_USAGE.getCode()) {
-
-        // resolve principal role granted to that principal
-        EntityCacheEntry principalRole =
-            this.resolveById(
-                toValidate,
-                PolarisEntityType.PRINCIPAL_ROLE,
-                PolarisEntityConstants.getNullId(),
-                grantRecord.getSecurableId());
-
-        // skip if purged or has been dropped
-        if (principalRole != null && !principalRole.getEntity().isDropped()) {
-          // add it to the activated list if no scoped principal role or this principal role is
-          // activated
-          if (callerPrincipalRoleNamesScope == null
-              || callerPrincipalRoleNamesScope.contains(principalRole.getEntity().getName())) {
-            // this principal role is activated
-            this.resolvedCallerPrincipalRoles.add(principalRole);
-          }
-        }
-      }
-    }
+    // activate all principal roles specified in the authenticated principal
+    resolvedCallerPrincipalRoles =
+        this.polarisPrincipal.getActivatedPrincipalRoleNames().isEmpty()
+            ? resolveAllPrincipalRoles(toValidate, resolvedCallerPrincipal)
+            : resolvePrincipalRolesByName(
+                toValidate, this.polarisPrincipal.getActivatedPrincipalRoleNames());
 
     // total success
     return new ResolverStatus(ResolverStatus.StatusEnum.SUCCESS);
   }
 
+  /**
+   * Resolve all principal roles that the principal has grants for
+   *
+   * @param toValidate
+   * @param resolvedCallerPrincipal1
+   * @return the list of resolved principal roles the principal has grants for
+   */
+  private List<EntityCacheEntry> resolveAllPrincipalRoles(
+      List<EntityCacheEntry> toValidate, EntityCacheEntry resolvedCallerPrincipal1) {
+    return resolvedCallerPrincipal1.getGrantRecordsAsGrantee().stream()
+        .filter(gr -> gr.getPrivilegeCode() == PolarisPrivilege.PRINCIPAL_ROLE_USAGE.getCode())
+        .map(
+            gr ->
+                resolveById(
+                    toValidate,
+                    PolarisEntityType.PRINCIPAL_ROLE,
+                    PolarisEntityConstants.getRootEntityId(),
+                    gr.getSecurableId()))
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Resolve the specified list of principal roles. The SecurityContext is used to determine whether
+   * the principal actually has the roles specified.
+   *
+   * @param toValidate
+   * @param roleNames
+   * @return the filtered list of resolved principal roles
+   */
+  private List<EntityCacheEntry> resolvePrincipalRolesByName(
+      List<EntityCacheEntry> toValidate, Set<String> roleNames) {
+    return roleNames.stream()
+        .filter(securityContext::isUserInRole)
```

Comment:
Why not use `AuthenticatedPolarisPrincipal` as the source of active role membership info? It looks like the Web App filters create both the principal and the security context, so the role information is available in the same logical scope as the principal.